### DISCUSSION TITLE
[Bug]: Training fails with dataloader

### DISCUSSION THREAD

**Discussion by stevejonathancitriot** [open-edge-platform/anomalib#1933] [Q&A]
### Describe the bug

I get an error when I try to access the dataloader items with the hazelnut_toy dataset

### Dataset

Folder

### Model

PatchCore

### Steps to reproduce the behavior

Installed anomalib with pip and "anomalib install"

Error when running this code

from anomalib.data import Folder

# Create the datamodule
datamodule = Folder(
    name="hazelnut_toy",
    root="/home/citriot/Downloads/hazelnut_toy",
    normal_dir="good",
    abnormal_dir="crack",
    mask_dir="mask/crack",
    normal_split_ratio=0.2,
)

# Setup the datamodule
datamodule.setup()

train_dataset = datamodule.train_data
val_dataset = datamodule.val_data
test_dataset = datamodule.test_data

# Access the dataloaders
train_dataloader = datamodule.train_dataloader()
val_dataloader = datamodule.val_dataloader()
test_dataloader = datamodule.test_dataloader()

i, train_data = next(enumerate(datamodule.train_dataloader()))
print(train_data.keys(), train_data['image'].shape)
# dict_keys(['image_path', 'label', 'image', 'mask_path', 'mask'])

i, val_data = next(enumerate(datamodule.val_dataloader()))
print(val_data.keys(), val_data['image'].shape)
# dict_keys(['image_path', 'label', 'image', 'mask_path', 'mask'])

i, test_data = next(enumerate(datamodule.test_dataloader()))
print(test_data.keys(), test_data['image'].shape)
# dict_keys(['image_path', 'label', 'image', 'mask_path', 'mask'])

from anomalib.models import Patchcore
from anomalib.engine import Engine

# Create the model and engine
model = Patchcore()
engine = Engine()

# Train a Patchcore model on the given datamodule
engine.train(datamodule=datamodule, model=model)

### OS information

OS information:
- OS: [e.g. Ubuntu 22.04]
- Python version: [e.g. 3.10.0]
- Anomalib version: [e.g. 1.0.0]
- PyTorch version: [e.g. 2.1.2+cu121]
- CUDA/cuDNN version: [e.g. 12.1]
- GPU models and configuration: [e.g. 1x GeForce RTX 3060]
- Any other relevant information: [e.g. I'm using a custom dataset]

### Expected behavior

Model would train?

### Pip/GitHub

pip

### What version/branch did you use?

1.0

### Configuration YAML

```yaml
/
```

### Logs

```shell
Traceback (most recent call last):
  File "/home/citriot/steve/anomalib/train_test.py", line 25, in <module>
    i, train_data = next(enumerate(datamodule.train_dataloader()))
  File "/home/citriot/miniconda3/envs/anomalibv2/lib/python3.10/site-packages/torch/utils/data/dataloader.py", line 630, in __next__
    data = self._next_data()
  File "/home/citriot/miniconda3/envs/anomalibv2/lib/python3.10/site-packages/torch/utils/data/dataloader.py", line 1345, in _next_data
    return self._process_data(data)
  File "/home/citriot/miniconda3/envs/anomalibv2/lib/python3.10/site-packages/torch/utils/data/dataloader.py", line 1371, in _process_data
    data.reraise()
  File "/home/citriot/miniconda3/envs/anomalibv2/lib/python3.10/site-packages/torch/_utils.py", line 694, in reraise
    raise exception
RuntimeError: Caught RuntimeError in DataLoader worker process 0.
Original Traceback (most recent call last):
  File "/home/citriot/miniconda3/envs/anomalibv2/lib/python3.10/site-packages/torch/utils/data/_utils/worker.py", line 308, in _worker_loop
    data = fetcher.fetch(index)
  File "/home/citriot/miniconda3/envs/anomalibv2/lib/python3.10/site-packages/torch/utils/data/_utils/fetch.py", line 54, in fetch
    return self.collate_fn(data)
  File "/home/citriot/miniconda3/envs/anomalibv2/lib/python3.10/site-packages/torch/utils/data/_utils/collate.py", line 265, in default_collate
    return collate(batch, collate_fn_map=default_collate_fn_map)
  File "/home/citriot/miniconda3/envs/anomalibv2/lib/python3.10/site-packages/torch/utils/data/_utils/collate.py", line 127, in collate
    return elem_type({key: collate([d[key] for d in batch], collate_fn_map=collate_fn_map) for key in elem})
  File "/home/citriot/miniconda3/envs/anomalibv2/lib/python3.10/site-packages/torch/utils/data/_utils/collate.py", line 127, in <dictcomp>
    return elem_type({key: collate([d[key] for d in batch], collate_fn_map=collate_fn_map) for key in elem})
  File "/home/citriot/miniconda3/envs/anomalibv2/lib/python3.10/site-packages/torch/utils/data/_utils/collate.py", line 123, in collate
    return collate_fn_map[collate_type](batch, collate_fn_map=collate_fn_map)
  File "/home/citriot/miniconda3/envs/anomalibv2/lib/python3.10/site-packages/torch/utils/data/_utils/collate.py", line 162, in collate_tensor_fn
    return torch.stack(batch, 0, out=out)
  File "/home/citriot/miniconda3/envs/anomalibv2/lib/python3.10/site-packages/torchvision/tv_tensors/_tv_tensor.py", line 77, in __torch_function__
    output = func(*args, **kwargs or dict())
RuntimeError: stack expects each tensor to be equal size, but got [3, 512, 512] at entry 0 and [3, 522, 512] at entry 14
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by JoannaCCJH** [comment]
Hi! I had similar issue before and solved it by adding `image_size=(256,256)` argument to `Folder()`
```python
datamodule = Folder(
  name="hazelnut_toy",
  root="/home/citriot/Downloads/hazelnut_toy",
  normal_dir="good",
  abnormal_dir="crack",
  mask_dir="mask/crack",
  normal_split_ratio=0.2,
  image_size=(256,256)
)
```

**Comment by stevejonathancitriot** [comment]
Sure that works, but with a custom dataset with the image_size parameter added, I'm still getting similar errors. I did try using torchvision transform to resize the images but that didn't work

**Comment by praycxy** [comment]
Have you solved it? I'm also having this issue and hope I can get your help

**Comment by samet-akcay** [comment]
> Sure that works, but with a custom dataset with the image_size parameter added, I'm still getting similar errors. I did try using torchvision transform to resize the images but that didn't work

@stevejonathancitriot, can you elaborate what sort of similar issues are you getting? I'm trying the reproduce the issue, but it works here

**Comment by samet-akcay** [comment]
When I try the setup you provided above, I can train and test the model

<img width="1222" alt="image" src="https://github.com/openvinotoolkit/anomalib/assets/17574420/7a240883-4f5e-4cb5-bc94-9fe1b40f6bde">

**Comment by samet-akcay** [comment]
This is because data modules expect transforms or image-size arguments. 
https://github.com/openvinotoolkit/anomalib/blob/6840780bcbb83d6197e3899a6161ff827f033314/src/anomalib/data/image/folder.py#L399-L402

When none of them is passed, the transforms are read from the model.
https://github.com/openvinotoolkit/anomalib/blob/6840780bcbb83d6197e3899a6161ff827f033314/src/anomalib/models/image/padim/lightning_model.py#L128-L136

If you instantiate the datamodule without these arguments and play with the datamodule to check the dataset items, you would get the error you shared. However, if you create a model and engine and train, it should work since the transforms will be obtained from model.

Converting this to a Q&A since this is not an issue
